File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-riskdef-5-4/PricingStructureValuation.cpp
// PricingStructureValuation.cpp 
#include "PricingStructureValuation.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

PricingStructureValuation::PricingStructureValuation(TiXmlNode* xmlNode)
: Valuation(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //baseDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* baseDateNode = xmlNode->FirstChildElement("baseDate");

   if(baseDateNode){baseDateIsNull_ = false;}
   else{baseDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- baseDateNode , address : " << baseDateNode << std::endl;
   #endif
   if(baseDateNode)
   {
      if(baseDateNode->Attribute("href") || baseDateNode->Attribute("id"))
      {
          if(baseDateNode->Attribute("id"))
          {
             baseDateIDRef_ = baseDateNode->Attribute("id");
             baseDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(baseDateNode));
             baseDate_->setID(baseDateIDRef_);
             IDManager::instance().setID(baseDateIDRef_,baseDate_);
          }
          else if(baseDateNode->Attribute("href")) { baseDateIDRef_ = baseDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { baseDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(baseDateNode));}
   }

   //spotDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* spotDateNode = xmlNode->FirstChildElement("spotDate");

   if(spotDateNode){spotDateIsNull_ = false;}
   else{spotDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- spotDateNode , address : " << spotDateNode << std::endl;
   #endif
   if(spotDateNode)
   {
      if(spotDateNode->Attribute("href") || spotDateNode->Attribute("id"))
      {
          if(spotDateNode->Attribute("id"))
          {
             spotDateIDRef_ = spotDateNode->Attribute("id");
             spotDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(spotDateNode));
             spotDate_->setID(spotDateIDRef_);
             IDManager::instance().setID(spotDateIDRef_,spotDate_);
          }
          else if(spotDateNode->Attribute("href")) { spotDateIDRef_ = spotDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { spotDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(spotDateNode));}
   }

   //inputDataDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* inputDataDateNode = xmlNode->FirstChildElement("inputDataDate");

   if(inputDataDateNode){inputDataDateIsNull_ = false;}
   else{inputDataDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- inputDataDateNode , address : " << inputDataDateNode << std::endl;
   #endif
   if(inputDataDateNode)
   {
      if(inputDataDateNode->Attribute("href") || inputDataDateNode->Attribute("id"))
      {
          if(inputDataDateNode->Attribute("id"))
          {
             inputDataDateIDRef_ = inputDataDateNode->Attribute("id");
             inputDataDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(inputDataDateNode));
             inputDataDate_->setID(inputDataDateIDRef_);
             IDManager::instance().setID(inputDataDateIDRef_,inputDataDate_);
          }
          else if(inputDataDateNode->Attribute("href")) { inputDataDateIDRef_ = inputDataDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { inputDataDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(inputDataDateNode));}
   }

   //endDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* endDateNode = xmlNode->FirstChildElement("endDate");

   if(endDateNode){endDateIsNull_ = false;}
   else{endDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- endDateNode , address : " << endDateNode << std::endl;
   #endif
   if(endDateNode)
   {
      if(endDateNode->Attribute("href") || endDateNode->Attribute("id"))
      {
          if(endDateNode->Attribute("id"))
          {
             endDateIDRef_ = endDateNode->Attribute("id");
             endDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(endDateNode));
             endDate_->setID(endDateIDRef_);
             IDManager::instance().setID(endDateIDRef_,endDate_);
          }
          else if(endDateNode->Attribute("href")) { endDateIDRef_ = endDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { endDate_ = boost::shared_ptr<IdentifiedDate>(new IdentifiedDate(endDateNode));}
   }

   //buildDateTimeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* buildDateTimeNode = xmlNode->FirstChildElement("buildDateTime");

   if(buildDateTimeNode){buildDateTimeIsNull_ = false;}
   else{buildDateTimeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- buildDateTimeNode , address : " << buildDateTimeNode << std::endl;
   #endif
   if(buildDateTimeNode)
   {
      if(buildDateTimeNode->Attribute("href") || buildDateTimeNode->Attribute("id"))
      {
          if(buildDateTimeNode->Attribute("id"))
          {
             buildDateTimeIDRef_ = buildDateTimeNode->Attribute("id");
             buildDateTime_ = boost::shared_ptr<XsdTypeDateTime>(new XsdTypeDateTime(buildDateTimeNode));
             buildDateTime_->setID(buildDateTimeIDRef_);
             IDManager::instance().setID(buildDateTimeIDRef_,buildDateTime_);
          }
          else if(buildDateTimeNode->Attribute("href")) { buildDateTimeIDRef_ = buildDateTimeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { buildDateTime_ = boost::shared_ptr<XsdTypeDateTime>(new XsdTypeDateTime(buildDateTimeNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<IdentifiedDate> PricingStructureValuation::getBaseDate()
{
   if(!this->baseDateIsNull_){
        if(baseDateIDRef_ != ""){
             return boost::shared_static_cast<IdentifiedDate>(IDManager::instance().getID(baseDateIDRef_));
        }else{
             return this->baseDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IdentifiedDate>();
   }
}
boost::shared_ptr<IdentifiedDate> PricingStructureValuation::getSpotDate()
{
   if(!this->spotDateIsNull_){
        if(spotDateIDRef_ != ""){
             return boost::shared_static_cast<IdentifiedDate>(IDManager::instance().getID(spotDateIDRef_));
        }else{
             return this->spotDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IdentifiedDate>();
   }
}
boost::shared_ptr<IdentifiedDate> PricingStructureValuation::getInputDataDate()
{
   if(!this->inputDataDateIsNull_){
        if(inputDataDateIDRef_ != ""){
             return boost::shared_static_cast<IdentifiedDate>(IDManager::instance().getID(inputDataDateIDRef_));
        }else{
             return this->inputDataDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IdentifiedDate>();
   }
}
boost::shared_ptr<IdentifiedDate> PricingStructureValuation::getEndDate()
{
   if(!this->endDateIsNull_){
        if(endDateIDRef_ != ""){
             return boost::shared_static_cast<IdentifiedDate>(IDManager::instance().getID(endDateIDRef_));
        }else{
             return this->endDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<IdentifiedDate>();
   }
}
boost::shared_ptr<XsdTypeDateTime> PricingStructureValuation::getBuildDateTime()
{
   if(!this->buildDateTimeIsNull_){
        if(buildDateTimeIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDateTime>(IDManager::instance().getID(buildDateTimeIDRef_));
        }else{
             return this->buildDateTime_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDateTime>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-riskdef-5-4/PricingStructureValuation.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class PricingStructureValuation : Valuation
    {
        public PricingStructureValuation(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode baseDateNode = xmlNode.SelectSingleNode("baseDate");
            
            if (baseDateNode != null)
            {
                if (baseDateNode.Attributes["href"] != null || baseDateNode.Attributes["id"] != null) 
                {
                    if (baseDateNode.Attributes["id"] != null) 
                    {
                        baseDateIDRef_ = baseDateNode.Attributes["id"].Value;
                        IdentifiedDate ob = new IdentifiedDate(baseDateNode);
                        IDManager.SetID(baseDateIDRef_, ob);
                    }
                    else if (baseDateNode.Attributes["href"] != null)
                    {
                        baseDateIDRef_ = baseDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        baseDate_ = new IdentifiedDate(baseDateNode);
                    }
                }
                else
                {
                    baseDate_ = new IdentifiedDate(baseDateNode);
                }
            }
            
        
            XmlNode spotDateNode = xmlNode.SelectSingleNode("spotDate");
            
            if (spotDateNode != null)
            {
                if (spotDateNode.Attributes["href"] != null || spotDateNode.Attributes["id"] != null) 
                {
                    if (spotDateNode.Attributes["id"] != null) 
                    {
                        spotDateIDRef_ = spotDateNode.Attributes["id"].Value;
                        IdentifiedDate ob = new IdentifiedDate(spotDateNode);
                        IDManager.SetID(spotDateIDRef_, ob);
                    }
                    else if (spotDateNode.Attributes["href"] != null)
                    {
                        spotDateIDRef_ = spotDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        spotDate_ = new IdentifiedDate(spotDateNode);
                    }
                }
                else
                {
                    spotDate_ = new IdentifiedDate(spotDateNode);
                }
            }
            
        
            XmlNode inputDataDateNode = xmlNode.SelectSingleNode("inputDataDate");
            
            if (inputDataDateNode != null)
            {
                if (inputDataDateNode.Attributes["href"] != null || inputDataDateNode.Attributes["id"] != null) 
                {
                    if (inputDataDateNode.Attributes["id"] != null) 
                    {
                        inputDataDateIDRef_ = inputDataDateNode.Attributes["id"].Value;
                        IdentifiedDate ob = new IdentifiedDate(inputDataDateNode);
                        IDManager.SetID(inputDataDateIDRef_, ob);
                    }
                    else if (inputDataDateNode.Attributes["href"] != null)
                    {
                        inputDataDateIDRef_ = inputDataDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        inputDataDate_ = new IdentifiedDate(inputDataDateNode);
                    }
                }
                else
                {
                    inputDataDate_ = new IdentifiedDate(inputDataDateNode);
                }
            }
            
        
            XmlNode endDateNode = xmlNode.SelectSingleNode("endDate");
            
            if (endDateNode != null)
            {
                if (endDateNode.Attributes["href"] != null || endDateNode.Attributes["id"] != null) 
                {
                    if (endDateNode.Attributes["id"] != null) 
                    {
                        endDateIDRef_ = endDateNode.Attributes["id"].Value;
                        IdentifiedDate ob = new IdentifiedDate(endDateNode);
                        IDManager.SetID(endDateIDRef_, ob);
                    }
                    else if (endDateNode.Attributes["href"] != null)
                    {
                        endDateIDRef_ = endDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        endDate_ = new IdentifiedDate(endDateNode);
                    }
                }
                else
                {
                    endDate_ = new IdentifiedDate(endDateNode);
                }
            }
            
        
            XmlNode buildDateTimeNode = xmlNode.SelectSingleNode("buildDateTime");
            
            if (buildDateTimeNode != null)
            {
                if (buildDateTimeNode.Attributes["href"] != null || buildDateTimeNode.Attributes["id"] != null) 
                {
                    if (buildDateTimeNode.Attributes["id"] != null) 
                    {
                        buildDateTimeIDRef_ = buildDateTimeNode.Attributes["id"].Value;
                        XsdTypeDateTime ob = new XsdTypeDateTime(buildDateTimeNode);
                        IDManager.SetID(buildDateTimeIDRef_, ob);
                    }
                    else if (buildDateTimeNode.Attributes["href"] != null)
                    {
                        buildDateTimeIDRef_ = buildDateTimeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        buildDateTime_ = new XsdTypeDateTime(buildDateTimeNode);
                    }
                }
                else
                {
                    buildDateTime_ = new XsdTypeDateTime(buildDateTimeNode);
                }
            }
            
        
        }
        
    
        #region BaseDate_
        private IdentifiedDate baseDate_;
        public IdentifiedDate BaseDate_
        {
            get
            {
                if (this.baseDate_ != null)
                {
                    return this.baseDate_; 
                }
                else if (this.baseDateIDRef_ != null)
                {
                    baseDate_ = IDManager.getID(baseDateIDRef_) as IdentifiedDate;
                    return this.baseDate_; 
                }
                else
                {
                      return this.baseDate_; 
                }
            }
            set
            {
                if (this.baseDate_ != value)
                {
                    this.baseDate_ = value;
                }
            }
        }
        #endregion
        
        public string baseDateIDRef_ { get; set; }
        #region SpotDate_
        private IdentifiedDate spotDate_;
        public IdentifiedDate SpotDate_
        {
            get
            {
                if (this.spotDate_ != null)
                {
                    return this.spotDate_; 
                }
                else if (this.spotDateIDRef_ != null)
                {
                    spotDate_ = IDManager.getID(spotDateIDRef_) as IdentifiedDate;
                    return this.spotDate_; 
                }
                else
                {
                      return this.spotDate_; 
                }
            }
            set
            {
                if (this.spotDate_ != value)
                {
                    this.spotDate_ = value;
                }
            }
        }
        #endregion
        
        public string spotDateIDRef_ { get; set; }
        #region InputDataDate_
        private IdentifiedDate inputDataDate_;
        public IdentifiedDate InputDataDate_
        {
            get
            {
                if (this.inputDataDate_ != null)
                {
                    return this.inputDataDate_; 
                }
                else if (this.inputDataDateIDRef_ != null)
                {
                    inputDataDate_ = IDManager.getID(inputDataDateIDRef_) as IdentifiedDate;
                    return this.inputDataDate_; 
                }
                else
                {
                      return this.inputDataDate_; 
                }
            }
            set
            {
                if (this.inputDataDate_ != value)
                {
                    this.inputDataDate_ = value;
                }
            }
        }
        #endregion
        
        public string inputDataDateIDRef_ { get; set; }
        #region EndDate_
        private IdentifiedDate endDate_;
        public IdentifiedDate EndDate_
        {
            get
            {
                if (this.endDate_ != null)
                {
                    return this.endDate_; 
                }
                else if (this.endDateIDRef_ != null)
                {
                    endDate_ = IDManager.getID(endDateIDRef_) as IdentifiedDate;
                    return this.endDate_; 
                }
                else
                {
                      return this.endDate_; 
                }
            }
            set
            {
                if (this.endDate_ != value)
                {
                    this.endDate_ = value;
                }
            }
        }
        #endregion
        
        public string endDateIDRef_ { get; set; }
        #region BuildDateTime_
        private XsdTypeDateTime buildDateTime_;
        public XsdTypeDateTime BuildDateTime_
        {
            get
            {
                if (this.buildDateTime_ != null)
                {
                    return this.buildDateTime_; 
                }
                else if (this.buildDateTimeIDRef_ != null)
                {
                    buildDateTime_ = IDManager.getID(buildDateTimeIDRef_) as XsdTypeDateTime;
                    return this.buildDateTime_; 
                }
                else
                {
                      return this.buildDateTime_; 
                }
            }
            set
            {
                if (this.buildDateTime_ != value)
                {
                    this.buildDateTime_ = value;
                }
            }
        }
        #endregion
        
        public string buildDateTimeIDRef_ { get; set; }
        
    
        
    
    }
    
}

